flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -0,0 +1,210 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.highavailability;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.HighAvailabilityOptions;
+import org.apache.flink.core.fs.FileStatus;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.jobmaster.JobResult;
+import org.apache.flink.runtime.rest.messages.json.JobResultDeserializer;
+import org.apache.flink.runtime.rest.messages.json.JobResultSerializer;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonIgnore;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Set;
+
+import static org.apache.flink.util.StringUtils.isNullOrWhitespaceOnly;
+
+/**
+ * An implementation of the {@link JobResultStore} which persists job result data to an underlying
+ * distributed filesystem.
+ */
+public class FileSystemJobResultStore extends AbstractThreadsafeJobResultStore {
+
+    private static final String DIRTY_SUFFIX = "_DIRTY";
+
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    private final FileSystem fileSystem;
+
+    private final Path basePath;
+
+    private final boolean deleteOnCommit;
+
+    @VisibleForTesting
+    FileSystemJobResultStore(FileSystem fileSystem, Path basePath, boolean deleteOnCommit)
+            throws IOException {
+        this.fileSystem = fileSystem;
+        this.basePath = basePath;
+        this.deleteOnCommit = deleteOnCommit;
+
+        this.fileSystem.mkdirs(this.basePath);
+    }
+
+    public static FileSystemJobResultStore fromConfiguration(Configuration config)
+            throws IOException {
+        Preconditions.checkNotNull(config);
+        final String jrsStoragePath = config.get(JobResultStoreOptions.STORAGE_PATH);
+        final Path basePath;
+
+        if (isNullOrWhitespaceOnly(jrsStoragePath)) {
+            final String haStoragePath = config.get(HighAvailabilityOptions.HA_STORAGE_PATH);
+            final String haClusterId = config.get(HighAvailabilityOptions.HA_CLUSTER_ID);
+            basePath = new Path(haStoragePath + "/job-result-store/" + haClusterId);
```

Comment:
I know that you found a generic solution for it in the documentation already. That's sufficient enough. I thought about it once more and just want to lay out a different approach (which you don't have to follow; just for future tasks). 

If you would have created a static method `FileSystemJobResultStore.createDefaultPath(String baseDirPath, String clusterId)` like:
```
public static String createDefaultJobResultStorePath(String baseDir, String clusterId) {
    return baseDir + "/job-result-store/" + clusterId;
}
```
which you could have used here
```suggestion
            basePath = new Path(createDefaultJobResultStorePath(haStoragePath, haClusterId));
```
you could have also used `FileSystemJobResultStore.createDefaultJobResultStore(String.format("{%s}", HighAvailabilityOptions.HA_STORAGE_PATH.key()), String.format("{%s}", HighAvailabilityOptions.HA_CLUSTER_ID.key())` in the description of `JobResultStoreOptions.STORAGE_PATH` to specify the default path without having the path multiple times in the codebase.

Reply:
This is a really good idea, I will implement this!